### PR TITLE
Fix heat scaling

### PR DESCRIPTION
**Pull request by guillaumehu**
@MUCDK noticed that the Chebyshev approximation was producing results different than another library. 

This PR fixes the scaling of the eigenvalue needed to compute the coefficients of the polynomials. There was also a problem with the eigenvalue of the rescaled Laplacian; we computed the coefficients with the eigenvalue prior to the rescaling. We now use the eigenvalue of the rescaled Laplacian (always equal to `2.0`) .

Now we also test if the approximated heat kernel is close to the ground truth one.

### CODE REVIEW

File: src/ott/geometry/geodesic.py [hunk 1]
```diff
@@ -116,12 +116,14 @@ def from_graph(
 
     if eigval is None:
       eigval = compute_largest_eigenvalue(laplacian, rng)
-    scaled_laplacian = jax.lax.cond((eigval > 2.0), lambda l: 2.0 * l / eigval,
-                                    lambda l: l, laplacian)
+
+    scaled_laplacian, eigval = jax.lax.cond((eigval > 2.0), lambda l:
+                                            (2.0 * l / eigval, 2.0), lambda l:
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:121]: Would use `0.5 * l`

File: src/ott/geometry/geodesic.py [hunk 2]
```diff
@@ -116,12 +116,14 @@ def from_graph(
...
+                                            (l, eigval), laplacian)
 
     # compute the coeffs of the Chebyshev pols approx using Bessel funcs
     chebyshev_coeffs = compute_chebychev_coeff_all(
-        eigval, t, order, laplacian.dtype
+        eigval / 2.0, t, order, laplacian.dtype
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:126]: Same here, `0.5 * eigval`

File: src/ott/geometry/geodesic.py [hunk 3]
```diff
@@ -149,7 +151,7 @@ def apply_kernel(
       Kernel applied to ``scaling``.
     """
     return expm_multiply(
-        self.scaled_laplacian, scaling, self.chebyshev_coeffs, self.eigval
+        self.scaled_laplacian, scaling, self.chebyshev_coeffs, self.eigval / 2.0
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:154]: 0.5 * self.eigval

File: tests/geometry/geodesic_test.py [hunk 4]
```diff
@@ -71,6 +71,24 @@ def gt_geometry(
   return geometry.Geometry(cost_matrix=cost, kernel_matrix=kernel, epsilon=1.)
 
 
+def exact_heat_kernel(
+    G: Union[jnp.ndarray, nx.Graph], normalize: bool = False, t: float = 10
```
Comment (michalk8) [inline tests/geometry/geodesic_test.py:75]: G cannot be `nx.Graph` here, but just an array, please fix.

File: tests/geometry/geodesic_test.py [hunk 5]
```diff
@@ -71,6 +71,24 @@ def gt_geometry(
...
+):
+  L = jnp.diag(jnp.sum(G, axis=1)) - G
+  if normalize:
+    inv_sqrt_deg = jnp.diag(
+        jnp.where(
+            jnp.sum(G, axis=1) > 0.0, 1.0 / jnp.sqrt(jnp.sum(G, axis=1)), 0.0
```
Comment (michalk8) [inline tests/geometry/geodesic_test.py:81]: Can you please reformat, so that's a bit nicer? Maybe
```python
out_deg = jnp.sum(G, axis=1)
inv_sqrt_deg = jnp.where(out_deg > 0.0, 1.0 / out_deg, 0.0)
...
```

File: tests/geometry/geodesic_test.py [hunk 6]
```diff
@@ -71,6 +71,24 @@ def gt_geometry(
...
+        )
+    )
+    L = inv_sqrt_deg @ L @ inv_sqrt_deg
+
+  e, v = jnp.linalg.eigh(L)
+  e = jnp.clip(e, 0, None)
```
Comment (michalk8) [inline tests/geometry/geodesic_test.py:87]: The `None` is not necessary here, please remove.

File: tests/geometry/geodesic_test.py [hunk 7]
```diff
@@ -99,16 +117,13 @@ def test_kernel_is_symmetric_positive_definite(
...
-    np.testing.assert_array_equal(vec_direct0, vec_direct1)
+    np.testing.assert_allclose(vec0, vec1, rtol=tol, atol=tol)
+    np.testing.assert_allclose(vec_direct0, vec_direct1, rtol=tol, atol=tol)
 
     np.testing.assert_allclose(vec0, vec_direct0, rtol=tol, atol=tol)
     np.testing.assert_allclose(vec1, vec_direct1, rtol=tol, atol=tol)
 
-    # compute the distance matrix and check that it is symmetric
     cost_matrix = geom.cost_matrix
-    np.testing.assert_array_equal(cost_matrix, cost_matrix.T)
```
Comment (michalk8) [inline tests/geometry/geodesic_test.py:110]: Why was this removed?

File: tests/geometry/geodesic_test.py [hunk 8]
```diff
@@ -240,3 +255,15 @@ def callback(geom: geodesic.Geodesic) -> float:
     expected = callback(geom__finite_right) - callback(geom__finite_left)
     actual = 2 * jnp.vdot(v_w, grad_sl)
     np.testing.assert_allclose(actual, expected, rtol=1e-4, atol=1e-4)
+
+  @pytest.mark.parametrize("normalize", [False, True])
+  def test_heat_approx(self, normalize: bool):
```
Comment (michalk8) [inline tests/geometry/geodesic_test.py:260]: I'd also parametrize by `t` and `order`.

**Review by michalk8 — CHANGES_REQUESTED** [review]
Thanks a lot @guillaumehu , just minor comments!

File: src/ott/geometry/geodesic.py (hunk 1 above)
Comment (michalk8) [reply src/ott/geometry/geodesic.py:121]: My mistake, this looks good!

**Review by michalk8 — APPROVED** [review]
Thanks @guillaumehu , merging this!

**Comment by guillaumehu** [comment]
Perfect, thank you for reviewing the PR!
